caffe2/operators/onnxifi_op.cc

```diff
@@ -0,0 +1,192 @@
+  for (const auto& s : ws_blobs) {
+    auto it = initialization_list->find(s);
+    if (it != initialization_list->end()) {
+      weight_names->emplace_back(s);
+      descs.emplace_back();
+      auto& tensor_desc = descs.back();
+      tensor_desc.name = weight_names->back().c_str();
+      BlobToTensorDescriptor(s, ws, &tensor_desc, weight_shapes);
+      initialization_list->erase(it);
+    }
+  }
+  CAFFE_ENFORCE(
+      initialization_list->empty(), "Unfulfilled initialization list");
+  return descs;
+}
+
+template <>
+bool OnnxifiOp<float, CPUContext>::RunOnDevice() {
+  for (unsigned i = 0U; i < InputSize(); ++i) {
+    const auto& input_tensor = Input(i);
+    const auto& tensor_dims = input_tensor.dims();
+    auto& tensor_descriptor = input_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    input_shapes_.emplace_back();
+    auto& input_shape = input_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      input_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = input_shape.data();
+    tensor_descriptor.buffer = (onnxPointer)(input_tensor.data<float>());
+  }
+
+  for (unsigned i = 0U; i < OutputSize(); ++i) {
+    auto* output_tensor = Output(i);
+    std::vector<TIndex> tensor_dims;
+    SetOutputShape(i, &tensor_dims);
+    output_tensor->Resize(tensor_dims);
+    auto& tensor_descriptor = output_desc_.at(i);
+    tensor_descriptor.dataType = ONNXIFI_DATATYPE_FLOAT32;
+    tensor_descriptor.memoryType = ONNXIFI_MEMORY_TYPE_CPU;
+    tensor_descriptor.dimensions = tensor_dims.size();
+    output_shapes_.emplace_back();
+    auto& output_shape = output_shapes_.back();
+    for (unsigned j = 0U; j < tensor_descriptor.dimensions; ++j) {
+      output_shape.push_back(tensor_dims[j]);
+    }
+    tensor_descriptor.shape = output_shape.data();
+    tensor_descriptor.buffer =
+        (onnxPointer)(output_tensor->mutable_data<float>());
+  }
+
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxSetGraphIO(
+          graph_,
+          input_desc_.size(),
+          input_desc_.data(),
+          output_desc_.size(),
+          output_desc_.data()),
+      ONNXIFI_STATUS_SUCCESS);
+
+  onnxMemoryFence input_fence;
+  input_fence.event = nullptr;
+  input_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxInitEvent(backend_, input_fence.event), ONNXIFI_STATUS_SUCCESS);
+  onnxMemoryFence output_fence;
+  output_fence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+  output_fence.event = nullptr;
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxInitEvent(backend_, output_fence.event), ONNXIFI_STATUS_SUCCESS);
+
+  // Singal event on input fence, call the asycn run on backend and wait for the event on output fence
+  CAFFE_ENFORCE_EQ(
+      lib_->onnxSignalEvent(input_fence.event), ONNXIFI_STATUS_SUCCESS);
```

Comment:
While this should work, I'd suggest to put `onnxSignalEvent` after `onnxRunGraph`, to check that backends handle this situation well.